django/contrib/auth/views.py

```diff
@@ -117,23 +119,38 @@ class LogoutView(SuccessURLAllowedHostsMixin, TemplateView):
     Log out the user and display the 'You are logged out' message.
     """
 
+    # RemovedInDjango50Warning: when the deprecation ends, remove "get" and
+    # "head" from http_method_names.
+    http_method_names = ["get", "head", "post", "options"]
     next_page = None
     redirect_field_name = REDIRECT_FIELD_NAME
     template_name = "registration/logged_out.html"
     extra_context = None
 
+    # RemovedInDjango50Warning: when the deprecation ends, move
+    # @method_decorator(csrf_protect) from post() to dispatch().
     @method_decorator(never_cache)
     def dispatch(self, request, *args, **kwargs):
+        if request.method.lower() == "get":
+            warnings.warn(
+                "Log out via GET requests is deprecated and will be removed in Django "
+                "5.0. Use POST requests for logging out.",
+                RemovedInDjango50Warning,
+            )
+        return super().dispatch(request, *args, **kwargs)
+
+    @method_decorator(csrf_protect)
+    def post(self, request, *args, **kwargs):
+        """Logout may be done via POST."""
         auth_logout(request)
         next_page = self.get_next_page()
         if next_page:
             # Redirect to this page until the session has been cleared.
             return HttpResponseRedirect(next_page)
-        return super().dispatch(request, *args, **kwargs)
+        return super().get(request, *args, **kwargs)
```

Comment:
This is funky and looks wrong, couldn't we just do the two lines here instead?
```
context = self.get_context_data(**kwargs)
return self.render_to_response(context)
```

Also I don't know if it's been mentioned but rendering during a successful POST shouldn't really be done, Django uses the [“post-redirect-get” pattern](https://en.wikipedia.org/wiki/Post/Redirect/Get) everywhere in normal forms to avoid refreshes causing repeat actions. Couldn't we apply that here? Perhaps complicating things a lot though...

Reply:
> This is funky and looks wrong, ...

Why? :thinking: Do you mean that `get()` shouldn't be called in `post()`? We use this pattern in many places and I don't see anything wrong in it.

> Django uses the “post-redirect-get” pattern everywhere in normal forms to avoid refreshes causing repeat actions. Couldn't we apply that here? Perhaps complicating things a lot though...

Is it not already applied when `next_page` is set?

Reply:
> Why? 🤔 Do you mean that get() shouldn't be called in post()? We use this pattern in many places and I don't see anything wrong in it.

Okay, maybe I'm not used to common flows in class-based views.

> Is it not already applied when next_page is set?

Yes it is applied in that case. Still, a user could:

1. Click "logout", see the "logged out" page
2. Open a new tab and login
3. Restart their browser
4. The browser resubmits the "logout" tab, and the user is logged out again.

I guess this pattern was here before, and this could be a separate issue.

Reply:
> The browser resubmits the "logout" tab, and the user is logged out again.

The browser asks whether it should resubmit POST requests no? But you are right it is not 100% nice if `next_page` is not used.

Reply:
Yes they do, but I think on restarting and being presented with confirm on top of a blank screen, most people would just click "yes, resubmit".

Reply:
> Yes they do, but I think on restarting and being presented with confirm on top of a blank screen, most people would just click "yes, resubmit".

Resubmit raises 403 in such case, so I don't think it's an issue:

https://user-images.githubusercontent.com/2865885/160332701-2a502657-ebe6-4a37-97d7-fa625856e9c9.mp4

Reply:
@adamchainz Does it work for you? :point_up:

Reply:
Ah, CSRF changing on logout saves us, of course. Brilliant. Thank you for testing! Not the kindest user experience but at least nothing breaks.